[ConstraintSystem] Try to be more efficient solving linked operator expressions

gregomni opened the pull request:
Explicitly check that every possible overload of the binary operator whose two arguments are the same type also results in the same type. (That is, for any `(T,T)` args `(T,T)->T`, with no `(T,T)->U`.)

Resolves [SR-10324](https://bugs.swift.org/browse/SR-10324).

The problem in that issue is that `(x * y) * z` requires the two multiplications to use two different overloads `(A,A)->B` in the parens and then `(B,A)->B` outside. But during constraint optimization, we see a series of binops with all the same type underlying variable types `(A,A,A)` and assume that we'll use the same overload for all of them (merging the type variables for both `*`s). Once we've done that, the constraint system is unsolvable.

This PR adds an additional check that none of the potential overloads has the form `(T,T)->U`, the existence of which is what breaks the optimization assumption.

Not done here, but I think that with this additional check it becomes safe to remove `isArithmeticOperatorDecl` and resolve this comment:
```
            // TODO: We currently limit this optimization to known arithmetic
            // operators, but we should be able to broaden this out to
            // logical operators as well.
```

gregomni committed:
Add condition for optimizing series of binops that every possible overload is (T,T)->T, with no (T,T)->U

gregomni commented:
@swift-ci Please smoke test.

xedin reviewed (dismissed):
I have attempted similar solution already since I'm tracking multiple dupes of this issue, the problem is that this makes some expressions too complex in our internal suites and I think it used to happen in public source compat suite as well :( 

I haven't had time since but if you really want to fix this, `LinkedExprAnalyzer` has to get removed from CSGen and its logic has to get ported over to the solver itself, this way we wouldn't have a problem with eagerness (and assumption that all of the overloads are homogeneous) and the same performance in the common case.

gregomni commented:
@xedin Ah, makes sense. Trying the public source-compat here, because it would be convenient to have a real code example to look at that fails on complexity, but then I'll go take a look at `LinkedExprAnalyzer` and see what I can do.

@swift-ci Please Test Source Compatibility

xedin commented:
I think you can add a custom overload to `+` and some of the tests in `validation-test/Sema/type_checker_perf` would fail. `import Foundation` and `import simd` make huge different too since they load a bunch of overloads to arithmetic operators.

gregomni committed:
Instead of chaining binops, favor disjunctions with op overloads whose types match existing binding choices

xedin commented:
@gregomni For inspiration, here is one of my attempts to get rid of `LinkedExprAnalyzer` and other hacks - https://github.com/apple/swift/pull/15246

gregomni committed:
Favor operators based on existing binds via both basename and type

gregomni committed:
Merge typevars of operators with shared decls after finding a solution to speed up further searching

gregomni committed:
Slight simplification of the operator merge logic

gregomni committed:
Dump disjunction possibilities on multiple aligned lines for better readability.

gregomni commented:
The couple days ago commits here essentially do as you (@xedin) suggest and moves the linked operator logic into the solver itself.

During overload selection on a disjunction, any overload for an operator which is either already bound in the constraint system (e.g. we're already using this exact `+`) or has identical types to an operator bound in the constraint system (e.g. we're using a `*` with these exact types) gets favored.

Also during selection of which disjunction gets picked next, it sorts on number of favored possibilities (if any) as well, so disjunctions with many possibilities but 1 (or a few) favored get explored before disjunctions with few possibilities over all but none favored.

With these changes the solver tends to search (approximately) depth-first down branches of matching operators.

Finally, when a solution is found with a given operator disjunction choice, all the typevars for any bindings in that solution to that exact overload are then merged (for examining any other possibilities). 

So the result is that the typevars aren't eagerly bound together, but the solver looks at same-overloads in operator expressions first, and if it is able to form a solution, the typevars are bound together afterwards to greatly cut down the remaining search space (identical to the original eager optimization, but delayed, and potentially piece-wise).

This works to fix sr-10324, the original goal here, but maintains almost all of the speed of the original optimization in the 'normal' arithmetical cases.

gregomni commented:
It turns out that "almost all of the speed" wasn't enough for `validation-test/Sema/type_checker_perf/fast/simd_add.gyb`, a long addition expression involving `SIMD4<Float>`s. Delaying the eager optimization was enough to break this test case because there are just so many overloads of `+` to look at before finding the solution and binding all occurrences together. So that got me looking at SIMD.

Today's commits here repurposes `LinkedExprAnalyzer`. The idea is to look at the known argument types and literals of the operators in the expression and to form a closed algebra containing all of the types that the intermediates and final result could be. (E.g. a simple example is if we're doing ordinary arithmetic involving integer literals, we could be dealing with any of the `Int` or `UInt` or `Float` types, but there's no way with any of those starting types and possible operators to get a `String` result, so `String` is outside of the domain & range of the algebra for the expression.)

If we're able to do that (form a closed domain of types), then we can disable all of the potential overloads for the operators that don't involve types in the domain and not explore them at all, because they are impossible. This is a very productive optimization when SIMD is involved. 

E.g. SR-10461:
```
% time xcrun swiftc -typecheck sr10461.swift                                                  
sr10461.swift:4:28: error: the compiler is unable to type-check this expression in reasonable time; try breaking up the expression into distinct sub-expressions
    let i = 3*(-p0 + 3*p1) - (3*p2 + p3)
            ~~~~~~~~~~~~~~~^~~~~~~~~~~~~
xcrun swiftc -typecheck sr10461.swift  16.92s user 0.44s system 99% cpu 17.409 total
```
vs
```
% time xcrun build/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/swiftc -typecheck sr10461.swift
xcrun build/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/swiftc -typecheck   8.95s user 0.30s system 99% cpu 9.293 total
```

It's still a big buggy with this code not correctly determining in a few circumstances that it can't prove the domain is actually closed. (And thus it disables overloads that it should leave alone.) Should work those out soon.

gregomni commented:
@swift-ci Please test compiler performance

gregomni commented:
This all works now. There are 4 validation tests that fail. They all include `-solver-enable-operator-designated-types`, and that setting overlaps and interferes somewhat with the disjunction handling added here. Removing the special `-solver` switches from those 4 tests cause them all to compile correctly, but since they are mostly scaling tests and I can't access the radars to tell what their purpose is, I'm not sure if or how I should modify those test files. They are:
```
validation-test/Sema/type_checker_perf/fast/rdar18360240.swift.gyb
validation-test/Sema/type_checker_perf/fast/rdar23327871.swift.gyb
validation-test/Sema/type_checker_perf/fast/rdar25866240.swift.gyb
validation-test/Sema/type_checker_perf/fast/rdar46687985.swift
```

gregomni commented:
@swift-ci Please test compiler performance

gregomni commented:
@swift-ci Please Test Source Compatibility

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -355,7 +365,233 @@ namespace {
-  
+
+  bool expandCollectedTypesFromOperatorOverloads(ConstraintSystem &CS,
+                                                 LinkedTypeInfo &lti) {
```
I'm actually hoping that we could remove all of this code from CSGen instead of adding more :)

xedin reviewed (commented):
Honestly after trying so many different things I'm starting to think that the only way to properly fix this would be attempting literal bindings earlier and backtracking if they don't match keeping a set of already attempted types. The first step is to remove `LinkedExprAnalyzer` completely and go from there... Another reason why we need to get rid of all of the perf hacks in CSGen is diagnostics, in "diagnostic mode" solver should be able to attempt all possible overloads and bindings to be able to find a solution.

The reason why we can't solve operators efficiently is related to the fact that overloads like `(Self, Self) -> Self` when opened create a single type variable for parameters and a result type, which makes the constraint system un-splittable into separate components. Possible ways to improve that I see is to either open `Self` as separate type variables in different positions and then have a separate solver step to `Type::join` them together (and possibly re-attempt some components with correct contextual type), or attempt bindings for arguments earlier, especially for the literals, because literal expressions have a type variable type with a literal protocol conformance which then forms `<T_param> arg conv <T_arg>` constraints so the graph could only be disconnected only if one of the sides gets a type.

gregomni commented:
@xedin I understand your hesitance about doing more pre-checking but I think narrowing the number of possibilities in large overload disjunctions is a productive way forward. I'm traveling the next few days, but I'd like to investigate the source compatibility failures here when I have the chance, and then try to convince you that this is a worthwhile thing to do once they are taken care of.

Meanwhile, if you think the commits on delaying operator tyvar merging (the first 5 up through `4ef4342604d79e5e19398fc8bcd1dbf015335974`) are worth contributing as a separate PR, I think those stand on their own.

Finally, I think a good future step would be to do this sort of thing in the solver itself rather than as a pre-step in CSGen. I imagine a separate possible-overloads disjunction constraint type, which references the tyvars for the operator arguments and attempts simplification by reducing the set of valid overloads as the argument tyvars get more constrained. In these arithmetic expressions it really seems like the problem is the large branching factor in constraint scopes, and reducing the fanout helps a lot.

xedin commented:
> I understand your hesitance about doing more pre-checking but I think narrowing the number of possibilities in large overload disjunctions is a productive way forward. 

That's what we keep trying to do and keep finding edge cases with. I think it's time for us to take a step back and try  to see if it could be done in the solver instead. @DougGregor was working on similar ideas, he might be able to share some of his thinking. @rudkx was investigating a different approach with designated operators.

gregomni commented:
@swift-ci Please Test Source Compatibility

gregomni commented:
@swift-ci Please Test Source Compatibility

gregomni commented:
Hmm. Getting identical failures when tried on the nightly toolchain (`swift-DEVELOPMENT-SNAPSHOT-2019-07-30-a.xctoolchain`) for Sourcery that _seems_ unrelated to my changes:
```
SIL verification failed: Unknown formal access pattern: storage
Verifying instruction:
   %7 = argument of bb1 : $*Optional<EJSTemplate.Error> // user: %95
->   %95 = begin_access [read] [dynamic] [no_nested_conflict] %7 : $*Optional<EJSTemplate.Error> // users: %96, %97
     %96 = load %95 : $*Optional<EJSTemplate.Error> // users: %131, %142, %105, %98
     end_access %95 : $*Optional<EJSTemplate.Error> // id: %97
```

gregomni commented:
@swift-ci Please smoke test

gregomni commented:
@swift-ci Please test source compatibility.

gregomni committed:
Extra work in picking disjunction not necessary.

gregomni commented:
@swift-ci Please test source compatibility.

gregomni commented:
@swift-ci Please smoke test.

gregomni commented:
Resolves: [SR-10324](https://bugs.swift.org/browse/SR-10324), [SR-10461](https://bugs.swift.org/browse/SR-10461), [SR-10601](https://bugs.swift.org/browse/SR-10601),  [SR-10606](https://bugs.swift.org/browse/SR-10606). 

So with this PR we'd no longer immediately bind arithmetic operators to be identical (SR-10324), but at the same time it greatly improves type checking time for expressions with lots of operators, especially those involving simd. (And passes source compatibility except for the current RxSwift `incompatible with weak references` error that ought to be an XFAIL.)

gregomni commented:
@swift-ci Please smoke test.

gregomni commented:
@swift-ci Please smoke test.

gregomni commented:
@swift-ci Please smoke test.

gregomni commented:
@swift-ci  Please smoke test.

gregomni commented:
@swift-ci Please test.

gregomni commented:
@swift-ci Please test Linux platform

gregomni commented:
@swift-ci Please smoke test.

gregomni commented:
@swift-ci Please test compiler performance.

gregomni commented:
@swift-ci Please test.

gregomni commented:
@swift-ci Please test compiler performance.

gregomni commented:
@swift-ci Please test compiler performance

gregomni commented:
@swift-ci Please test compiler performance

gregomni commented:
Help! Any ideas what this failure testing compiler performance means or what to do about it?
```
19:41:54 WORKSPACE: /Users/buildnode/jenkins/workspace/swift-PR-compiler-performance-macOS
19:41:54 Build step 'Execute shell' marked build as failure
19:41:54 Archiving artifacts
19:53:19 [BFA] Scanning build for known causes...
19:53:26 [BFA] No failure causes found
19:53:26 [BFA] Done. 7s
```

xedin commented:
@shahmishal ^

gregomni commented:
Got some conflicts that needed resolving, so let's try this again.

@swift-ci Please test compiler performance

gregomni commented:
@swift-ci Please test compiler performance

gregomni commented:
@shahmishal Is the script for 'test compiler performance' available? I'm not finding it, but I'm not sure where to look. I can't tell whether the error here is a failure of `process-stats-dir.py`, or of whatever the next step is after that.

gregomni commented:
Actually, looking at other PRs, being marked as failing and including the `Build step 'Execute shell' marked build as failure` line, and then going on and producing results anyway seems to be common, so my guess that that line was the issue appears wrong, and I'm back to step zero. :)

shahmishal commented:
You want to look at git@github.com:apple/swift-source-compat-suite.git, this is the repository that has the compiler performance script.

gregomni commented:
thanks!

gregomni commented:
# Summary for master full

**Unexpected test results, excluded stats for Core, XcodeEdit, RxCocoa, Cryptor, GRDB, SnapshotTesting, Proverb, Wordy, SwifterSwift**

**Regressions found (see below)**

- [Debug-batch](#debug-batch-100452be243a6e4d)
    - [debug-batch brief](#debug-batch-brief-100452be243a6e4d)
    - [debug-batch detailed](#debug-batch-detailed-100452be243a6e4d)
- [Release](#release-100452be243a6e4d)
    - [release brief](#release-brief-100452be243a6e4d)
    - [release detailed](#release-detailed-100452be243a6e4d)



<a name="debug-batch-100452be243a6e4d">

# Debug-batch

<a name="debug-batch-brief-100452be243a6e4d">

## debug-batch brief
<details open>
<summary>Regressed (1)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
time.swift-driver.wall | 2987.5s | 3851.8s | 864.2s | 28.93% :no_entry:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (2)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 35,802,402,959,522 | 35,547,976,236,371 | -254,426,723,151 | -0.71%
LLVM.NumLLVMBytesOutput | 1,700,610,102 | 1,699,351,180 | -1,258,922 | -0.07%
</details>

<a name="debug-batch-detailed-100452be243a6e4d">

## debug-batch detailed
<details>
<summary>Regressed (6)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Sema.ExtendedNominalRequest | 3,830,211 | 3,947,059 | 116,848 | 3.05% :no_entry:
Sema.InheritedDeclsReferencedRequest | 4,631,269 | 4,819,404 | 188,135 | 4.06% :no_entry:
Sema.NumConformancesDeserialized | 7,419,740 | 7,690,421 | 270,681 | 3.65% :no_entry:
Sema.NumLazyIterableDeclContexts | 6,906,305 | 7,040,338 | 134,033 | 1.94% :no_entry:
Sema.NumUnloadedLazyIterableDeclContexts | 4,220,844 | 4,360,901 | 140,057 | 3.32% :no_entry:
Sema.SelfBoundsFromWhereClauseRequest | 6,151,648 | 6,416,797 | 265,149 | 4.31% :no_entry:
</details>
<details>
<summary>Improved (7)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumSourceLinesPerSecond | 4,563,706 | 4,396,821 | -166,885 | -3.66% :white_check_mark:
Driver.NumDriverPipePolls | 88,581 | 85,035 | -3,546 | -4.0% :white_check_mark:
Driver.NumDriverPipeReads | 80,265 | 76,803 | -3,462 | -4.31% :white_check_mark:
Sema.NamedLazyMemberLoadSuccessCount | 23,818,077 | 23,295,031 | -523,046 | -2.2% :white_check_mark:
Sema.NumConstraintScopes | 22,209,964 | 16,549,360 | -5,660,604 | -25.49% :white_check_mark:
Sema.NumConstraintsConsideredForEdgeContraction | 57,666,080 | 39,533,352 | -18,132,728 | -31.44% :white_check_mark:
Sema.NumLeafScopes | 14,567,248 | 10,935,579 | -3,631,669 | -24.93% :white_check_mark:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (132)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumASTBytesAllocated | 47,940,824,355 | 47,836,627,120 | -104,197,235 | -0.22%
AST.NumDecls | 133,757 | 133,658 | -99 | -0.07%
AST.NumDependencies | 192,464 | 192,351 | -113 | -0.06%
AST.NumInfixOperators | 51,091 | 51,041 | -50 | -0.1%
AST.NumLinkLibraries | 0 | 0 | 0 | 0.0%
AST.NumLoadedModules | 249,917 | 249,660 | -257 | -0.1%
AST.NumLocalTypeDecls | 250 | 250 | 0 | 0.0%
AST.NumObjCMethods | 24,678 | 24,675 | -3 | -0.01%
AST.NumPostfixOperators | 23 | 23 | 0 | 0.0%
AST.NumPrecedenceGroups | 24,585 | 24,556 | -29 | -0.12%
AST.NumPrefixOperators | 84 | 84 | 0 | 0.0%
AST.NumReferencedDynamicNames | 189 | 189 | 0 | 0.0%
AST.NumReferencedMemberNames | 6,017,073 | 6,012,172 | -4,901 | -0.08%
AST.NumReferencedTopLevelNames | 457,100 | 456,636 | -464 | -0.1%
AST.NumSourceBuffers | 288,429 | 288,098 | -331 | -0.11%
AST.NumSourceLines | 4,488,978 | 4,486,463 | -2,515 | -0.06%
AST.NumTotalClangImportedEntities | 4,649,829 | 4,646,548 | -3,281 | -0.07%
Driver.ChildrenMaxRSS | 207,268,407,296 | 205,838,909,440 | -1,429,497,856 | -0.69%
Driver.DriverDepCascadingDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingTopLevel | 0 | 0 | 0 | 0.0%
Driver.DriverDepDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepTopLevel | 0 | 0 | 0 | 0.0%
Driver.NumDriverJobsRun | 26,564 | 26,523 | -41 | -0.15%
Driver.NumDriverJobsSkipped | 0 | 0 | 0 | 0.0%
Driver.NumProcessFailures | 0 | 0 | 0 | 0.0%
Frontend.MaxMallocUsage | 879,812,244,656 | 877,272,064,840 | -2,540,179,816 | -0.29%
Frontend.NumInstructionsExecuted | 35,802,402,959,522 | 35,547,976,236,371 | -254,426,723,151 | -0.71%
Frontend.NumProcessFailures | 0 | 0 | 0 | 0.0%
IRModule.NumIRAliases | 191,517 | 191,336 | -181 | -0.09%
IRModule.NumIRBasicBlocks | 6,529,196 | 6,527,055 | -2,141 | -0.03%
IRModule.NumIRComdatSymbols | 0 | 0 | 0 | 0.0%
IRModule.NumIRFunctions | 3,219,521 | 3,216,645 | -2,876 | -0.09%
IRModule.NumIRGlobals | 3,376,802 | 3,374,446 | -2,356 | -0.07%
IRModule.NumIRIFuncs | 0 | 0 | 0 | 0.0%
IRModule.NumIRInsts | 83,349,845 | 83,296,749 | -53,096 | -0.06%
IRModule.NumIRNamedMetaData | 127,922 | 127,787 | -135 | -0.11%
IRModule.NumIRValueSymbols | 5,978,327 | 5,973,602 | -4,725 | -0.08%
LLVM.NumLLVMBytesOutput | 1,700,610,102 | 1,699,351,180 | -1,258,922 | -0.07%
Parse.NumFunctionsParsed | 256,546 | 256,442 | -104 | -0.04%
Parse.NumIterableDeclContextParsed | 558,836 | 558,554 | -282 | -0.05%
SILModule.NumSILGenDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILGenFunctions | 1,635,562 | 1,634,149 | -1,413 | -0.09%
SILModule.NumSILGenGlobalVariables | 51,178 | 51,092 | -86 | -0.17%
SILModule.NumSILGenVtables | 18,202 | 18,198 | -4 | -0.02%
SILModule.NumSILGenWitnessTables | 67,839 | 67,789 | -50 | -0.07%
SILModule.NumSILOptDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILOptFunctions | 2,337,779 | 2,335,846 | -1,933 | -0.08%
SILModule.NumSILOptGlobalVariables | 52,710 | 52,621 | -89 | -0.17%
SILModule.NumSILOptVtables | 30,620 | 30,610 | -10 | -0.03%
SILModule.NumSILOptWitnessTables | 148,203 | 148,077 | -126 | -0.09%
Sema.AccessLevelRequest | 9,967,629 | 9,908,475 | -59,154 | -0.59%
Sema.AttachedFunctionBuilderRequest | 7 | 7 | 0 | 0.0%
Sema.AttachedPropertyWrapperTypeRequest | 493,488 | 492,682 | -806 | -0.16%
Sema.AttachedPropertyWrappersRequest | 2,399,870 | 2,400,009 | 139 | 0.01%
Sema.CollectOverriddenDeclsRequest | 5,822,123 | 5,801,246 | -20,877 | -0.36%
Sema.CursorInfoRequest | 0 | 0 | 0 | 0.0%
Sema.CustomAttrNominalRequest | 17 | 17 | 0 | 0.0%
Sema.DefaultAndMaxAccessLevelRequest | 76,548 | 76,511 | -37 | -0.05%
Sema.DefaultDefinitionTypeRequest | 7,471 | 7,470 | -1 | -0.01%
Sema.DefaultTypeRequest | 431,230 | 431,621 | 391 | 0.09%
Sema.EmittedMembersRequest | 25,278 | 25,265 | -13 | -0.05%
Sema.EnumRawTypeRequest | 20,630 | 20,619 | -11 | -0.05%
Sema.ExistentialConformsToSelfRequest | 17,146 | 17,146 | 0 | 0.0%
Sema.ExistentialTypeSupportedRequest | 14,952 | 14,946 | -6 | -0.04%
Sema.FunctionBuilderTypeRequest | 7 | 7 | 0 | 0.0%
Sema.GetDestructorRequest | 25,778 | 25,770 | -8 | -0.03%
Sema.HasDynamicMemberLookupAttributeRequest | 0 | 0 | 0 | 0.0%
Sema.InheritedTypeRequest | 280,932 | 281,108 | 176 | 0.06%
Sema.InitKindRequest | 98,608 | 98,499 | -109 | -0.11%
Sema.IsAccessorTransparentRequest | 298,316 | 298,085 | -231 | -0.08%
Sema.IsDeclApplicableRequest | 0 | 0 | 0 | 0.0%
Sema.IsDynamicRequest | 1,523,510 | 1,522,224 | -1,286 | -0.08%
Sema.IsFinalRequest | 4,145,639 | 4,140,986 | -4,653 | -0.11%
Sema.IsGetterMutatingRequest | 407,297 | 406,796 | -501 | -0.12%
Sema.IsImplicitlyUnwrappedOptionalRequest | 2,263,135 | 2,262,662 | -473 | -0.02%
Sema.IsObjCRequest | 1,450,943 | 1,448,701 | -2,242 | -0.15%
Sema.IsSetterMutatingRequest | 329,992 | 329,593 | -399 | -0.12%
Sema.LazyStoragePropertyRequest | 2,065 | 2,065 | 0 | 0.0%
Sema.MangleLocalTypeDeclRequest | 500 | 499 | -1 | -0.2%
Sema.NamedLazyMemberLoadFailureCount | 16,884 | 16,882 | -2 | -0.01%
Sema.NominalTypeLookupDirectCount | 32,603,191 | 32,583,105 | -20,086 | -0.06%
Sema.NumAccessorBodiesSynthesized | 181,554 | 181,378 | -176 | -0.1%
Sema.NumAccessorsSynthesized | 268,024 | 267,752 | -272 | -0.1%
Sema.NumCyclicOneWayComponentsCollapsed | 0 | 0 | 0 | 0.0%
Sema.NumDeclsDeserialized | 57,013,494 | 57,109,772 | 96,278 | 0.17%
Sema.NumDeclsTypechecked | 1,347,263 | 1,346,030 | -1,233 | -0.09%
Sema.NumDeclsValidated | 2,317,946 | 2,315,768 | -2,178 | -0.09%
Sema.NumFunctionsTypechecked | 506,127 | 505,819 | -308 | -0.06%
Sema.NumGenericSignatureBuilders | 1,137,017 | 1,135,330 | -1,687 | -0.15%
Sema.NumLazyGenericEnvironments | 10,552,769 | 10,527,393 | -25,376 | -0.24%
Sema.NumLazyGenericEnvironmentsLoaded | 287,231 | 287,944 | 713 | 0.25%
Sema.NumLazyRequirementSignatures | 693,380 | 695,455 | 2,075 | 0.3%
Sema.NumLazyRequirementSignaturesLoaded | 465,674 | 465,223 | -451 | -0.1%
Sema.NumTypesDeserialized | 16,991,541 | 17,002,343 | 10,802 | 0.06%
Sema.NumTypesValidated | 1,635,440 | 1,634,170 | -1,270 | -0.08%
Sema.OpaqueReadOwnershipRequest | 254,103 | 253,831 | -272 | -0.11%
Sema.OverriddenDeclsRequest | 2,082,644 | 2,079,066 | -3,578 | -0.17%
Sema.PropertyWrapperBackingPropertyInfoRequest | 488,971 | 488,237 | -734 | -0.15%
Sema.PropertyWrapperBackingPropertyTypeRequest | 493,488 | 492,682 | -806 | -0.16%
Sema.PropertyWrapperMutabilityRequest | 577,206 | 576,350 | -856 | -0.15%
Sema.PropertyWrapperTypeInfoRequest | 5 | 5 | 0 | 0.0%
Sema.ProtocolRequiresClassRequest | 66,849 | 67,040 | 191 | 0.29%
Sema.ProvideDefaultImplForRequest | 5,822,123 | 5,801,246 | -20,877 | -0.36%
Sema.RangeInfoRequest | 0 | 0 | 0 | 0.0%
Sema.RequirementRequest | 102,599 | 102,596 | -3 | -0.0%
Sema.RequirementSignatureRequest | 536,916 | 536,638 | -278 | -0.05%
Sema.RequiresOpaqueAccessorsRequest | 1,243,922 | 1,242,871 | -1,051 | -0.08%
Sema.RequiresOpaqueModifyCoroutineRequest | 243,545 | 243,337 | -208 | -0.09%
Sema.ResilienceExpansionRequest | 1,597,277 | 1,596,198 | -1,079 | -0.07%
Sema.ResolveProtocolNameRequest | 0 | 0 | 0 | 0.0%
Sema.RootAndResultTypeOfKeypathDynamicMemberRequest | 0 | 0 | 0 | 0.0%
Sema.RootTypeOfKeypathDynamicMemberRequest | 0 | 0 | 0 | 0.0%
Sema.SelfAccessKindRequest | 4,576,178 | 4,572,857 | -3,321 | -0.07%
Sema.SetterAccessLevelRequest | 140,691 | 140,545 | -146 | -0.1%
Sema.StorageImplInfoRequest | 1,113,274 | 1,111,982 | -1,292 | -0.12%
Sema.StoredPropertiesAndMissingMembersRequest | 29,113 | 29,092 | -21 | -0.07%
Sema.StoredPropertiesRequest | 270,985 | 270,654 | -331 | -0.12%
Sema.StructuralTypeRequest | 0 | 0 | 0 | 0.0%
Sema.SuperclassDeclRequest | 363,663 | 363,689 | 26 | 0.01%
Sema.SuperclassTypeRequest | 47,131 | 47,116 | -15 | -0.03%
Sema.SynthesizeAccessorRequest | 268,024 | 267,752 | -272 | -0.1%
Sema.TypeCheckFunctionBodyUntilRequest | 506,127 | 505,819 | -308 | -0.06%
Sema.TypeDeclsFromWhereClauseRequest | 43,621 | 43,600 | -21 | -0.05%
Sema.TypeRelationCheckRequest | 0 | 0 | 0 | 0.0%
Sema.USRGenerationRequest | 6,825,065 | 6,809,166 | -15,899 | -0.23%
Sema.UnderlyingTypeDeclsReferencedRequest | 213,345 | 214,021 | 676 | 0.32%
</details>

<a name="release-100452be243a6e4d">

# Release

<a name="release-brief-100452be243a6e4d">

## release brief
<details open>
<summary>Regressed (1)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
time.swift-driver.wall | 4854.3s | 5499.2s | 644.9s | 13.29% :no_entry:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (2)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 42,639,709,190,067 | 42,386,851,514,978 | -252,857,675,089 | -0.59%
LLVM.NumLLVMBytesOutput | 1,491,466,106 | 1,491,518,474 | 52,368 | 0.0%
</details>

<a name="release-detailed-100452be243a6e4d">

## release detailed
<details>
<summary>Regressed (3)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Sema.NumConformancesDeserialized | 3,721,763 | 3,802,966 | 81,203 | 2.18% :no_entry:
Sema.NumLazyGenericEnvironmentsLoaded | 33,955 | 34,381 | 426 | 1.25% :no_entry:
Sema.NumLazyIterableDeclContexts | 1,267,946 | 1,301,258 | 33,312 | 2.63% :no_entry:
</details>
<details>
<summary>Improved (1)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Sema.NumConstraintScopes | 21,975,598 | 16,237,124 | -5,738,474 | -26.11% :white_check_mark:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (17)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumLoadedModules | 28,760 | 28,760 | 0 | 0.0%
AST.NumTotalClangImportedEntities | 1,173,363 | 1,173,018 | -345 | -0.03%
IRModule.NumIRBasicBlocks | 5,332,725 | 5,333,408 | 683 | 0.01%
IRModule.NumIRFunctions | 2,732,734 | 2,732,733 | -1 | -0.0%
IRModule.NumIRGlobals | 2,922,032 | 2,921,992 | -40 | -0.0%
IRModule.NumIRInsts | 50,417,697 | 50,421,278 | 3,581 | 0.01%
IRModule.NumIRValueSymbols | 5,308,662 | 5,308,621 | -41 | -0.0%
LLVM.NumLLVMBytesOutput | 1,491,466,106 | 1,491,518,474 | 52,368 | 0.0%
SILModule.NumSILGenFunctions | 1,139,215 | 1,139,265 | 50 | 0.0%
SILModule.NumSILOptFunctions | 1,594,922 | 1,597,575 | 2,653 | 0.17%
Sema.NumDeclsDeserialized | 9,884,711 | 9,927,257 | 42,546 | 0.43%
Sema.NumDeclsValidated | 1,753,921 | 1,753,921 | 0 | 0.0%
Sema.NumFunctionsTypechecked | 510,381 | 510,381 | 0 | 0.0%
Sema.NumGenericSignatureBuilders | 289,391 | 289,270 | -121 | -0.04%
Sema.NumLazyGenericEnvironments | 1,955,729 | 1,958,098 | 2,369 | 0.12%
Sema.NumTypesDeserialized | 5,300,846 | 5,316,705 | 15,859 | 0.3%
Sema.NumTypesValidated | 936,268 | 936,268 | 0 | 0.0%
</details>

gregomni commented:
`run_cperf` works locally (results just above), so I'm not sure what's up with CI on this PR.

gregomni commented:
@swift-ci Please test source compatibility

gregomni commented:
@swift-ci Please test

gregomni commented:
The wall time is presumably unreliable with this last performance test because it was running on my local machine (see the couple percent improvements in the previous runs by CI without significant -  if any - algorithmic changes after).

So this is a set of changes that, yes, unfortunately adds about 150 lines of code as a pre-check in CSGen. But there was already an existing pre-checking walk happening there already, so the additional complication is not that much.

In exchange, you get to drop the early-forced-identical operator bindings hack, and there are a full *one quarter* fewer solver scopes! Which means many more arithmetic-esque expressions solvable in shorter times. 

Please review?

xedin commented:
I will put some time aside to take a look at this upcoming week.

xedin commented:
Sorry @gregomni I have this open and ready just need to finish up some other stuff first before I dig in again.

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -356,7 +389,183 @@ namespace {
+    }
+
+    for (auto convertTy : lti.collectedTypes) {
+      if (CS.TC.isConvertibleTo(convertTy, type, CS.DC))
```
This `isConvertiableTo` (which creates new constraint system every time) and `conformsToProtocol` are pretty heavy checks which, in my opinion, we are better off only performing while solving for each individual overload. 

There is a long standing improvement opportunity where `determineBestBindings` could actually check whether found type conforms to all of the protocols associated with its type variable via `{Literal}ConformsTo` constraint and avoid that binding.

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -356,7 +389,183 @@ namespace {
+
+          bool possibleDecl = true;
+
+          if (auto generic = resultTy->getAs<GenericTypeParamType>()) {
```
Just a note - overloads like `(Self, Self) -> Self` or `(T, T) -> T` are the biggest problem is at the moment - inability to filter out generic overloads which connect arguments to a result and prevent efficient solving of disconnected components, which means we have to explore a lot deeper into solution trees.

xedin replied:
This could be simplified down to 

```c++
if (llvm::any_of(fnType->getParams(), [&resultTy](const AnyFunctionType::Param &param) {
   return resultTy->isEqual(param.getPlainType());
})
  continue;
```

gregomni replied:
> Just a note - overloads like `(Self, Self) -> Self` or `(T, T) -> T` are the biggest problem is at the moment - inability to filter out generic overloads which connect arguments to a result and prevent efficient solving of disconnected components, which means we have to explore a lot deeper into solution trees.

These types of overloads can't be ruled out as possibilities up front, nor disconnected, but this PR (by favoring choosing duplicate bindings of existing bindings and by merging those type variables once a solution exists) makes it so that the solution tree gets pruned to be much less branchy once we find a solution. The depth is unaffected, but the breadth gets vastly narrower.

xedin replied:
Sorry but I don't understand what you are trying to say, the problem is that in the worse case, especially when expression is invalid e.g `1 + "hello"`, we'd still get to explore almost everything just because generic overloads like `(Self, Self) -> Self` carry almost no context for the bindings and we can't verify conformances until generic parameter is actually bound to something.

gregomni replied:
Well, with this PR, you'd disable every possible overload of `+` except for the generic `(Self, Self) -> Self` style, because all the non-generic ones are impossible. So you explore much less. I was trying to say that this is a big performance win, even when you can't do anything efficient with the generic style ones directly.

xedin replied:
What I'm trying to say is that presence of overloads with concrete types is not a problem if arguments are concrete as well or could be bound to something relatively easily.

xedin replied:
Like in the examples you have in the PR, all of the arguments are going to be concrete so it's pretty easy for the solver to avoid going deeper if argument conversions could be tested and rejected right away.

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -356,7 +389,183 @@ namespace {
+        continue;
+
+      llvm::TinyPtrVector<Constraint *> disjunctions =
+          CS.getConstraintGraph().gatherConstraints(
```
Performance wise - it would be better to invert that check and have outer loop be for every disjunction constraint, at least it's going to be linear over a number of constraints in the system.

xedin reviewed (commented):
I see the benefit but I'm still skeptical about this approach. The problem for me, as I mentioned before, is the fact that merging of type variables associated with literal arguments shouldn't happen during constraint generation. If we can avoid that it would spare us a lot of work which is currently done upfront e.g. trying to collect all possible types for literal arguments, disabling overloads.

As an alternative approach - while solving, we could try to re-arrange choices in other disjunctions to much the previous choice taken for the same operator e.g. `1 + 2 + 3` binding first `+` to `(Int, Int) -> Int` would bubble up `(Int, Int) -> Int` to the top so it could be attempted first for the second `+` as well. 

It seems such behavior could imitate this (LinkedExprAnalyzer) optimization almost exactly for "concrete" overloads because there is another "hack" in the solver which doesn't attempt generic overloads when there is a concrete match already available.

gregomni commented:
@xedin
> I see the benefit but I'm still skeptical about this approach. The problem for me, as I mentioned before, is the fact that merging of type variables associated with literal arguments shouldn't happen during constraint generation. If we can avoid that it would spare us a lot of work which is currently done upfront e.g. trying to collect all possible types for literal arguments, disabling overloads.

Sorry, I'm not sure I understand this comment. Can you expand, please? There isn't any merging of type variables happening during constraint generation here. (The old -- sr-10324 bug causing -- merging of operator overload type variables is removed as part of this PR.)

This does examine all the literals in order to determine which operator overloads can be disabled as impossible in this expression, but the literal type variables aren't touched and can still potentially be solved as different types from each other.

> As an alternative approach - while solving, we could try to re-arrange choices in other disjunctions to much the previous choice taken for the same operator e.g. `1 + 2 + 3` binding first `+` to `(Int, Int) -> Int` would bubble up `(Int, Int) -> Int` to the top so it could be attempted first for the second `+` as well.

This is part of what is already in this PR. If we've chosen `(Int, Int) -> Int` for the first binding, then the `(Int, Int) -> Int` binding becomes preferred when looking at the second overload disjunction.

xedin commented:
> This does examine all the literals in order to determine which operator overloads can be disabled as impossible in this expression, but the literal type variables aren't touched and can still potentially be solved as different types from each other.

Sorry I was referring to the old behavior which was problematic. I understand that new logic is trying to disable some overloads upfront, but I just don't think that CSGen is a good place for that since it could be done while solving without the overhead of examining overload sets.

> This is part of what is already in this PR. If we've chosen (Int, Int) -> Int for the first binding, then the (Int, Int) -> Int binding becomes preferred when looking at the second overload disjunction.

Favoring overloads but itself is not enough, I think to get the biggest win here and be 1-1 with previous behavior solver would have to - get newly favored overload to the top of the list, and adjust `selectBestBindingDisjunction` to pick the same kind of operator next if it's adjacent. That way we can quickly test the same overload for all adjacent operators in the chain and stop if it matches. I think with that strategy we can remove `LinkedExprAnalyzer` completely.

gregomni commented:
> Sorry I was referring to the old behavior which was problematic. I understand that new logic is trying to disable some overloads upfront, but I just don't think that CSGen is a good place for that since it could be done while solving without the overhead of examining overload sets.

Well, there are two pieces, there is (1) determining the subset of all types that the expression could possibly involve, and then there is (2) avoiding taking paths in the solution tree that involve impossible types. 

(1) is a property of the expression as a whole, not any particular variable or constraint, and so I don't see a reasonable way to do it during solving. It seems clearly to belong during CSGen to me. On the other hand, it would be possible to have the constraint system hold the set of possible types and do (2), the disjunction path filtering, entirely during solving. But the code for determining which overloads are skippable given the set of possible types is very similar to determining the set of types in the first place, and so sharing code and containing it and the set of types in one place seems better to me.

I think of it as generating a less bushy tree in the first place, which seems a reasonable job for CSGen to be doing.

> Favoring overloads but itself is not enough, I think to get the biggest win here and be 1-1 with previous behavior solver would have to - get newly favored overload to the top of the list, and adjust `selectBestBindingDisjunction` to pick the same kind of operator next if it's adjacent. That way we can quickly test the same overload for all adjacent operators in the chain and stop if it matches. I think with that strategy we can remove `LinkedExprAnalyzer` completely.

I did effectively that, favoring the matching overload (which in these sorts of expressions is almost always the only favored choice, then), and sorting disjunctions by minimum-favored. The solver immediately picked all the same operators.

It turns out that this was not enough to solve several of the SIMD expressions in tests once the old `LinkedExprAnalyzer` overload type-var merging was removed. So there were several test failures taking too much time to solve.

It also turns out that, once the new `LinkedExprAnalyzer` here was in, the extra work to pick an optimum disjunction order was a net performance loss, so I removed it in e18f1355f1c6267e53ab2e17b83180704ffdafd1.

xedin commented:
Here is how I see this - (1) is impossible to do properly in CSGen in presence of generic overloads, because it was possible we could use a classic algorithm with deterministic type variable ordering and solver wouldn't have any exponential behavior; (2) Solver is already pretty good at not taking any path where both sides of the argument conversion constraint or result are concrete, so pruning not a problem in this case, but when it comes to generic overloads all bets off at the moment. 

> It turns out that this was not enough to solve several of the SIMD expressions in tests once the old LinkedExprAnalyzer overload type-var merging was removed. So there were several test failures taking too much time to solve.

This is exactly the main problem we are yet to solve which no hacks or tweaks to thereof can. I think we should concentrate on trying to solve a problem related to over-exploration in case of generic overloads, once that is done we could remove all of the existing hacks as obsolete.

One of the ways to fix this problem for literal arguments - is to attempt default literal types earlier and backtrack if they don't match.

gregomni committed:
Fix for API change in simplifyLocatorToAnchor()

gregomni commented:
Did the cleanup you suggested, rebased to HEAD, fixed for a change in the `simplifyLocatorToAnchor()` API.

> Here is how I see this - (1) is impossible to do properly in CSGen in presence of generic overloads, because it was possible we could use a classic algorithm with deterministic type variable ordering and solver wouldn't have any exponential behavior; (2) Solver is already pretty good at not taking any path where both sides of the argument conversion constraint or result are concrete, so pruning not a problem in this case, but when it comes to generic overloads all bets off at the moment.

Determining subset of types (1) is actually trivial (but unhelpful) for generic overloads. If it takes `Self` and returns `Self` then the possible types are exactly the same as they would be in the absence of the operator altogether. So it's true that these overloads are never disabled, but they are easy to deal with in the algorithm.

As for (2): I think both the scope performance counter and the list of bugs which the solver previously failed to solve and with this PR successfully solves shows that the solver can still use some help here.

> This is exactly the main problem we are yet to solve which no hacks or tweaks to thereof can. I think we should concentrate on trying to solve a problem related to over-exploration in case of generic overloads, once that is done we could remove all of the existing hacks as obsolete.

I understand that this PR does not solve what you see as the main problem. It greatly ameliorates the situation by improving performance and solving expressions that weren't getting solved otherwise. It also replaces a hack that had incorrect language semantics - erroring in the case of valid code (SR-10324) - with (what you see as) a different hack that doesn't have that problem.

I don't think we know how to solve the main problem right now. Wouldn't it be better to successfully compile more valid code until we figure it out?

gregomni commented:
@swift-ci Please smoke test.

xedin commented:
> I don't think we know how to solve the main problem right now. Wouldn't it be better to successfully compile more valid code until we figure it out?

The reason why I'm hesitant to take these changes is as follows - every time we try to tweak CSGen/solver hacks we get either more "too complex" reports or source breaks. I can't prove to myself that this is not going to cause either. At this point I'm happy to maintain the status quo until we can up with principled solution which doesn't require any hacks even though some of the valid code is going to stay broken meanwhile.

gregomni committed:
Remove LinkedExprAnalyzer completely. Short-circuit generic overload disjunctions by testing last opened typevar.

gregomni committed:
scale-test can't fit a curve to these now, converted to simple tests at max size.

gregomni commented:
@swift-ci Please test source compatibility.

futurejones commented:
Can one of the admins verify this patch?

CodaFi commented:
@gregomni I'm going to close this out due to age and inactivity. Thank you for taking a look here.
